joblib/numpy_pickle.py

```diff
@@ -270,6 +273,28 @@ def save(self, obj):
                         traceback.format_exc()))
         return Pickler.save(self, obj)
 
+    def save_bytes(self, obj):
+        """Strongly inspired from python 2.7 pickle.Pickle.save_bytes"""
+        if self.bin:
+            n = len(obj)
+            if n < 256:
+                self.write(pickle.SHORT_BINSTRING + asbytes(chr(n)) + obj)
+            else:
+                self.write(pickle.BINSTRING + struct.pack("<i", n) + obj)
```

Comment:
Not much difference either.

Minor detail: in order to trigger the single-file pickle code, the joblib.dump `cache_size` argument has to be set (defaults to 100MB).

```python

import joblib
from joblib import numpy_pickle

import numpy as np

a = np.zeros(int(999 * 1024 ** 2), dtype=np.uint8)  # 999 MB
%load_ext memory_profiler
%mprun -f numpy_pickle.NumpyPickler.save_bytes joblib.dump(a, '/tmp/a.pkl', compress=1, cache_size=1024)
```

### Single write with string concatenation
```
Line #    Mem usage    Increment   Line Contents
================================================
   276   1030.0 MiB      0.0 MiB       def save_bytes(self, obj):
   277                                     """Strongly inspired from python 2.7 pickle.Pickler.save_string"""
   278   1030.0 MiB      0.0 MiB           if self.bin:
   279   1030.0 MiB      0.0 MiB               n = len(obj)
   280   1030.0 MiB      0.0 MiB               if n < 256:
   281   1030.0 MiB      0.0 MiB                   self.write(pickle.SHORT_BINSTRING + asbytes(chr(n)) + obj)
   282                                         else:
   283   2029.1 MiB    999.1 MiB                   self.write(pickle.BINSTRING + struct.pack("<i", n) + obj)
   284                                             # self.write(pickle.BINSTRING)
   285                                             # self.write(struct.pack("<i", n))
   286                                             # self.write(obj)
   287   2029.1 MiB      0.0 MiB               self.memoize(obj)
   288                                     else:
   289                                         Pickler.save_bytes(self, obj)
```

### Multiple writes 
```
Line #    Mem usage    Increment   Line Contents
================================================
   276   1030.1 MiB      0.0 MiB       def save_bytes(self, obj):
   277                                     """Strongly inspired from python 2.7 pickle.Pickler.save_string"""
   278   1030.1 MiB      0.0 MiB           if self.bin:
   279   1030.1 MiB      0.0 MiB               n = len(obj)
   280   1030.1 MiB      0.0 MiB               if n < 256:
   281   1030.1 MiB      0.0 MiB                   self.write(pickle.SHORT_BINSTRING + asbytes(chr(n)) + obj)
   282                                         else:
   283                                             # self.write(pickle.BINSTRING + struct.pack("<i", n) + obj)
   284   1030.1 MiB      0.0 MiB                   self.write(pickle.BINSTRING)
   285   1030.1 MiB      0.0 MiB                   self.write(struct.pack("<i", n))
   286   2029.1 MiB    998.9 MiB                   self.write(obj)
   287   2029.2 MiB      0.1 MiB               self.memoize(obj)
   288                                     else:
   289                                         Pickler.save_bytes(self, obj)
```